skdownscale/pointwise_models/gard.py

```diff
@@ -121,11 +165,34 @@ def _predict_one_step(self, lr_model, X):
         x = np.asarray(self.kdtree_.data)[inds]
         y = self.y_[inds]
 
-        # train linear regression model
-        lr_model.fit(x, y)
+        # figure out if there's a threshold of interest
+        if self.thresh is not None:
+            exceed_ind = y > self.thresh
+        else:
+            exceed_ind = np.array([True] * len(y))
+
+        # train logistic regression model
+        if self.thresh is not None:
+            binary_y = exceed_ind.astype(int)
+            logistic_model.fit(x, binary_y)
+            exceedance_prob = logistic_model.predict_proba(X)[:, 0]
+            self.stats_['prob'][i] = exceedance_prob[0]
+
+        # train linear regression model otherwise
+        lr_model.fit(x[exceed_ind], y[exceed_ind])
+
+        # calculate the rmse of prediction
+        y_hat = lr_model.predict(x[exceed_ind])
+        error = mean_squared_error(y[exceed_ind], y_hat, squared=False)
+        self.stats_['error'][i] = error
 
-        # predict for this time step
         predicted = lr_model.predict(X)
+
+        if return_exceedance_prob:
+            return exceedance_prob
+        if return_errors:
+            return error
```

Comment:
I've been meaning to look at how sklearn would handle this. Presumably, they have patterns for this sort of thing. Have you found anything there that would allow us to return the predictions, probabilities, and/or errors at the same time?

Reply:
Or perhaps retrieve these quantities after prediction?

Something like:
```python
predicted = model.predict(X)
probs = model.get_probs()
error = model.get_error()
```

Reply:
Good question -- I believe in sklearn these would simply be different APIs (e.g. `.predict` vs `.predict_proba` vs. `.score`), and X/y are provided to these APIs each time. For the implementation here I was trying to work with a couple constraints: 

1. the Pointwise Wrapper currently only has .fit, .predict, and .transform APIs. So if we implement these in any other function or store them as attributes, we'll end up needing to iterate through each model in the Pointwise wrapper to access these. Instead of implementing everything in .predict with different arguments to circumvent this, perhaps we could implement a generic API in the wrapper to call any function/retrieve any attribute of the underlying model? 

2. the AnalogRegression model can be slow in predicting since it's iterating through each day. Thus it'd be great to avoid repeating computation if possible. This is the rationale for saving these stats in an attribute instead of recomputing with the provided X each time, but is of course prone to user error. Fixing the API as mentioned above would make this point moot.  

Thoughts?

Reply:
synced offline and determined to: 
1. save the probability and error as attributes during predict 
2. access these values with an get_attrs wrapper function in PointwiseDownscaler 

note that this breaks the sklearn pattern of [not changing the attributes of models during predict](https://github.com/scikit-learn/scikit-learn/blob/432ae47beeaf611547aafb69131834f151fe1136/sklearn/utils/estimator_checks.py#L1129-L1166), but making the compromise here to avoid re-run